packages/ui5-test-writer/src/index.ts

```diff
@@ -0,0 +1,209 @@
+        startupTargets = [startupTargets];
+    }
+
+    // Create page configurations in supported cases
+    const appTargets = manifest['sap.ui5']?.routing?.targets;
+    for (const targetKey in appTargets) {
+        const pageConfig = createPageConfig(manifest, targetKey);
+        if (pageConfig) {
+            pageConfig.isStartup = startupTargets.includes(targetKey);
+            config.pages.push(pageConfig);
+        }
+    }
+
+    return config;
+}
+
+/**
+ * Writes a page object in a mem-fs-editor.
+ *
+ * @param pageConfig - the page configuration object
+ * @param rootTemplateDirPath - template root directory
+ * @param testOutDirPath - output test directory (.../webapp/test)
+ * @param fs - a reference to a mem-fs editor
+ */
+function writePageObject(
+    pageConfig: FEV4OPAPageConfig,
+    rootTemplateDirPath: string,
+    testOutDirPath: string,
+    fs: Editor
+) {
+    fs.copyTpl(
+        join(rootTemplateDirPath, `integration/pages/${pageConfig.template}.js`),
+        join(testOutDirPath, `integration/pages/${pageConfig.targetKey}.js`),
+        pageConfig,
+        undefined,
+        {
+            globOptions: { dot: true }
+        }
+    );
+}
+
+/**
+ * Generate OPA test files for a Fiori elements for OData V4 application.
+ * Note: this can potentially overwrite existing files in the webapp/test folder.
+ *
+ * @param basePath - the absolute target path where the application will be generated
+ * @param opaConfig - parameters for the generation
+ * @param opaConfig.scriptName - the name of the OPA journey file. If not specified, 'FirstJourney' will be used
+ * @param fs - an optional reference to a mem-fs editor
+ * @returns Reference to a mem-fs-editor
+ */
+export function generateOPAFiles(basePath: string, opaConfig: { scriptName?: string }, fs?: Editor): Editor {
+    if (!fs) {
+        fs = create(createStorage());
+    }
+
+    const manifest = fs.readJSON(join(basePath, 'webapp/manifest.json')) as any as Manifest;
+    const config = createConfig(manifest, opaConfig);
+
+    const rootCommonTemplateDirPath = join(__dirname, '../templates/common');
+    const rootV4TemplateDirPath = join(__dirname, '../templates/v4'); // Only v4 is supported for the time being
+    const testOutDirPath = join(basePath, 'webapp/test');
+
+    // Test files
+    fs.copy(join(rootCommonTemplateDirPath, 'testsuite.qunit.html'), join(testOutDirPath, 'testsuite.qunit.html'));
+    fs.copy(join(rootCommonTemplateDirPath, 'testsuite.qunit.js'), join(testOutDirPath, 'testsuite.qunit.js'));
+
+    // Integration (OPA) test files
+    fs.copyTpl(
+        join(rootV4TemplateDirPath, 'integration', 'opaTests.*.*'),
+        join(testOutDirPath, 'integration'),
+        config,
+        undefined,
+        {
+            globOptions: { dot: true }
+        }
+    );
+
+    // Pages files (one for each page in the app)
+    config.pages.forEach((page) => {
+        writePageObject(page, rootV4TemplateDirPath, testOutDirPath, fs!);
```

Comment:
```suggestion
        writePageObject(page, rootV4TemplateDirPath, testOutDirPath, fs);
```

Shouldn't have to use the non-null assertion operator.

Instead of creating an editor and assigning to `fs` if it's not passed in, something like this would be preferable:
```typescript
    const editor = fs || create(createStorage());

   // Now use `editor` instead of `fs` in the code
```
We do re-assign to `fs` at other places in the repo but it's not good practice to re-assign function parameters.

Reply:
done